pgmpy/metrics/metrics.py

```diff
@@ -467,3 +467,189 @@ def SHD(true_model, est_model):
+          - A bidirected link u↔v is encoded by having both arcs u→v and v→u.
+
+        References
+        ----------
+        [1] Faller, P. M., et al. (2024).
+        “Self‐compatibility: Evaluating Causal Discovery without Ground Truth.”
+        In AISTATS. arXiv:2307.09552
+        """
+        # 1) Prep
+        full_nx = dag.to_directed()
+        V = set(dag.nodes())
+        latent = V - set(observed)
+
+        directed_edges = set()
+        bidirected_edges = set()
+
+        # 2) Preserve any original observed→observed arcs
+        for u, v in dag.edges():
+            if u in observed and v in observed:
+                directed_edges.add((u, v))
+
+        # 3) Detect latent‐only directed chains via induced subgraph
+        for u, v in combinations(observed, 2):
+            sub_nodes = latent | {u, v}
+            subG = full_nx.subgraph(sub_nodes)
+            if nx.has_path(subG, u, v):
+                directed_edges.add((u, v))
+            if nx.has_path(subG, v, u):
+                directed_edges.add((v, u))
+
+        # 4) Detect collider‐mediated connections → bidirected edges
+        #    Use active_trail_nodes to find d-connected partners including latents.
+        #    If v shows up in u’s active set *only* via a collider at some latent,
+        #    then that implies u↔v in the ADMG.
+        atn = dag.active_trail_nodes(
+            variables=list(observed),
+            observed=None,  # no observed conditioning
+            include_latents=True,  # so we see latent colliders too
+        )
+        for u in observed:
+            # every v ≠ u that’s d-connected to u
+            for v in atn[u] & set(observed) - {u}:
+                # skip if we already have a directed link one way or the other
+                if (u, v) in directed_edges or (v, u) in directed_edges:
+                    continue
+                # else this must be a collider‐only connection → bidirected
+                bidirected_edges.add((u, v))
+                bidirected_edges.add((v, u))
+
+        # 5) Build output
+        admg = nx.DiGraph()
+        admg.add_nodes_from(observed)
+        admg.add_edges_from(directed_edges)
+        admg.add_edges_from(bidirected_edges)
+        return admg
+
+    # 1) Fit joint model on all variables
+    joint_learner = estimator_class(data)
+    joint = joint_learner.estimate(**estimator_kwargs)
+    if not isinstance(joint, DAG):
+        joint = DiscreteBayesianNetwork(list(joint.edges()))
+
+    shd_values = []
+    # 2) For each random subset...
+    for i in range(num_subsets):
+        # a) sample subset of columns
+        S = rng.choice(variables, size=k, replace=False).tolist()
+        sub_data = data[S]
+        # b) fit marginal model
+        marg_learner = estimator_class(sub_data)
+        marginal = marg_learner.estimate(**estimator_kwargs)
+        # c) project joint onto S
+        joint_proj = _latent_admg(joint, S)
```

Comment:
Would save a lot of compute if we put this line outside the for loop as `joint_proj` would be the same in each loop.

Reply:
My understanding is that we compare the full structure on all the variables with a bunch of structures learned on subsets of variables. So, the full structure remains constant in each of these comparisons, so we can move it outside of the for loop.